ethstorage/p2p/discovery.go

```diff
@@ -114,15 +113,38 @@ func (conf *Config) Discovery(log log.Logger, l1ChainID uint64, tcpPort uint16,
 	}
 	udpV5, err := discover.ListenV5(conn, localNode, cfg)
 	if err != nil {
-		return nil, nil, err
+		return nil, nil, isIPSet, err
 	}
 
 	log.Info("Started discovery service", "enr", localNode.Node(), "id", localNode.ID(), "port", udpV5.Self().UDP())
 
 	// TODO: periodically we can pull the external IP and TCP port from libp2p NAT service,
 	// and add it as a statement to keep the localNode accurate (if we trust the NAT device more than the discv5 statements)
 
-	return localNode, udpV5, nil
+	return localNode, udpV5, isIPSet, nil
+}
+
+func updateLocalNodeIPAndTCP(addrs []ma.Multiaddr, localNode *enode.LocalNode) bool {
+	for _, addr := range addrs {
+		ipStr, err := addr.ValueForProtocol(4)
+		if err != nil {
+			continue
+		}
+		ip := net.ParseIP(ipStr)
+		if ip.IsPrivate() || !ip.IsGlobalUnicast() {
```

Comment:
Maybe we can reuse getLocalPublicIPv4?

Reply:
this function is to find out public IP from the node addresses (observed peers) for private node which is not equal to local public IP.